File: web/app/tests/Auth/AuthMethods.test.ts
import { resolveAuthMethods } from "@base/core"
import { describe, expect, it } from "vitest"

// Parity contract: the iOS (AuthMethodsTests.swift) and Android
// (AuthMethodsTest.kt) ports assert these same cases, so the three platforms
// resolve an AUTH_METHODS value identically.
describe("resolveAuthMethods", () => {
    it("defaults to email-code when nothing is configured", () => {
        expect(resolveAuthMethods({})).toEqual(["email-code"])
        expect(resolveAuthMethods({ methodsValue: "" })).toEqual(["email-code"])
        expect(resolveAuthMethods({ methodsValue: "  ,  " })).toEqual(["email-code"])
    })

    it("ignores unknown names and falls back when nothing valid remains", () => {
        expect(resolveAuthMethods({ methodsValue: "magic-beans, sms" })).toEqual(["email-code"])
        expect(resolveAuthMethods({ methodsValue: "magic-beans, password" })).toEqual(["password"])
    })

    it("resolves every registered provider key", () => {
        expect(
            resolveAuthMethods({
                methodsValue: "email-code,password,google,apple,github,facebook,discord,x,linkedin,anonymous",
            }),
        ).toEqual([
            "email-code",
            "password",
            "google",
            "apple",
            "github",
            "facebook",
            "discord",
            "x",
            "linkedin",
            "anonymous",
        ])
    })

    it("preserves configured order and dedupes", () => {
        expect(resolveAuthMethods({ methodsValue: "google, email-code, google" })).toEqual([
            "google",
            "email-code",
        ])
        expect(resolveAuthMethods({ methodsValue: "password,email-code,anonymous" })).toEqual([
            "password",
            "email-code",
            "anonymous",
        ])
        expect(resolveAuthMethods({ methodsValue: "apple, google, email-code" })).toEqual([
            "apple",
            "google",
            "email-code",
        ])
    })

    it("tolerates whitespace and case", () => {
        expect(resolveAuthMethods({ methodsValue: " Google , EMAIL-CODE " })).toEqual([
            "google",
            "email-code",
        ])
        expect(resolveAuthMethods({ methodsValue: " Apple " })).toEqual(["apple"])
    })

    it("appends google for the legacy flag without duplicating", () => {
        expect(resolveAuthMethods({ googleEnabled: true })).toEqual(["email-code", "google"])
        expect(resolveAuthMethods({ methodsValue: "google", googleEnabled: true })).toEqual(["google"])
    })
})

File: android/app/src/test/kotlin/com/baseapp/android/AuthMethodsTest.kt
package com.baseapp.android

import com.baseapp.android.auth.AuthMethod
import com.baseapp.android.auth.resolveAuthMethods
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Parity tests for resolveAuthMethods against web/core's AuthMethods.ts:
 * trim, lowercase, ignore unknown names, dedupe, preserve configured order,
 * default to email codes.
 */
class AuthMethodsTest {
    @Test
    fun defaultsToEmailCodeWhenValueIsNull() {
        assertEquals(listOf(AuthMethod.EMAIL_CODE), resolveAuthMethods(null))
    }

    @Test
    fun defaultsToEmailCodeWhenValueIsEmpty() {
        assertEquals(listOf(AuthMethod.EMAIL_CODE), resolveAuthMethods(""))
        assertEquals(listOf(AuthMethod.EMAIL_CODE), resolveAuthMethods("   "))
    }

    @Test
    fun defaultsToEmailCodeWhenValueIsGarbage() {
        assertEquals(listOf(AuthMethod.EMAIL_CODE), resolveAuthMethods("saml, ldap, ,,magic"))
    }

    @Test
    fun preservesConfiguredOrder() {
        assertEquals(
            listOf(AuthMethod.GOOGLE, AuthMethod.EMAIL_CODE),
            resolveAuthMethods("google, email-code"),
        )
        assertEquals(
            listOf(AuthMethod.PASSWORD, AuthMethod.GOOGLE, AuthMethod.ANONYMOUS),
            resolveAuthMethods("password,google,anonymous"),
        )
        assertEquals(
            listOf(AuthMethod.APPLE, AuthMethod.GOOGLE, AuthMethod.EMAIL_CODE),
            resolveAuthMethods("apple, google, email-code"),
        )
    }

    @Test
    fun dedupesRepeatedMethods() {
        assertEquals(
            listOf(AuthMethod.GOOGLE, AuthMethod.EMAIL_CODE),
            resolveAuthMethods("google, email-code, google, email-code"),
        )
    }

    @Test
    fun ignoresUnknownNamesButKeepsKnownOnes() {
        assertEquals(
            listOf(AuthMethod.PASSWORD, AuthMethod.GOOGLE),
            resolveAuthMethods("password, saml, google, magic-link"),
        )
    }

    @Test
    fun toleratesWhitespaceAndMixedCase() {
        assertEquals(
            listOf(AuthMethod.EMAIL_CODE, AuthMethod.GOOGLE),
            resolveAuthMethods("  Email-Code ,  GOOGLE  "),
        )
        assertEquals(listOf(AuthMethod.APPLE), resolveAuthMethods(" Apple "))
    }

    @Test
    fun resolvesEveryKnownMethodKey() {
        assertEquals(
            AuthMethod.entries.toList(),
            resolveAuthMethods(
                "email-code,password,google,apple,github,facebook,discord,x,linkedin,anonymous",
            ),
        )
    }

    @Test
    fun oauthProviderClassificationMirrorsTheRegistry() {
        assertEquals(
            listOf(
                AuthMethod.GOOGLE,
                AuthMethod.APPLE,
                AuthMethod.GITHUB,
                AuthMethod.FACEBOOK,
                AuthMethod.DISCORD,
                AuthMethod.X,
                AuthMethod.LINKEDIN,
            ),
            AuthMethod.entries.filter { it.isOAuthProvider },
        )
    }
}
